### ISSUE TITLE
bosh task is queued but never started

### ISSUE THREAD

**Issue by gossion** [cloudfoundry/bosh#2043]
**Describe the bug**
We met the issue several times occasionally in our test environment when deploying a new CF. The problem is the deployment stuck there and never start. From the log (please see section `Logs`), it shows that the task 7 was queued. Hours later, the status was still `queued`.

**To Reproduce**
Steps to reproduce the behavior (example):
1. Deploy a bosh director on Azure with manifests https://github.com/cloudfoundry/bosh-deployment/blob/13c473ea96d191864f37984d934598c7239c63ab/bosh.yml
2. Deploy a CF, which will try to create release at the beginning 

**Expected behavior**
Expect the deployment to start.

**Logs**
```
azureuser@devbox-8-28-7-34-11:~$ bosh -e azure tasks -r
Using environment '10.0.0.4' as client 'admin'

ID  State   Started At                    Last Activity At              User   Deployment  Description      Result
7   queued  Thu Jan  1 00:00:00 UTC 1970  Tue Aug 28 15:08:54 UTC 2018  admin  -           create release   -
6   done    Tue Aug 28 15:09:17 UTC 2018  Tue Aug 28 15:09:37 UTC 2018  admin  -           create release   Created release 'bpm/0.9.0'
5   done    Tue Aug 28 15:08:25 UTC 2018  Tue Aug 28 15:09:18 UTC 2018  admin  -           create release   Created release 'silk/2.10.0'
4   done    Tue Aug 28 15:09:18 UTC 2018  Tue Aug 28 15:09:21 UTC 2018  admin  -           create release   Created release 'bosh-dns-aliases/0.0.2'
3   done    Tue Aug 28 15:08:25 UTC 2018  Tue Aug 28 15:09:17 UTC 2018  admin  -           create release   Created release 'loggregator-agent/2.0'
2   done    Tue Aug 28 15:04:57 UTC 2018  Tue Aug 28 15:17:19 UTC 2018  admin  -           create stemcell  /stemcells/bosh-azure-hyperv-ubuntu-trusty-go_agent/3586.36
1   done    Tue Aug 28 15:04:39 UTC 2018  Tue Aug 28 15:04:56 UTC 2018  admin  -           create release   Created release 'bosh-dns/1.7.0'

7 tasks

Succeeded
azureuser@devbox-8-28-7-34-11:~$ bosh -e azure task 7 --debug
Using environment '10.0.0.4' as client 'admin'

Task 7

I, [2018-08-28T15:09:59.835121 #64] [0x2ae778134838]  INFO -- TaskHelper: Director Version: 266.7.0
I, [2018-08-28T15:09:59.835232 #64] [0x2ae778134838]  INFO -- TaskHelper: Enqueuing task: 7
^C

```

**Versions (please complete the following information):**
 - Infrastructure: "Azure"
 - BOSH version: "266.7.0"
 - BOSH CLI version: "version 2.0.48-e94aeeb-2018-01-09T23:08:07Z"
 - Stemcell version: "https://bosh.io/d/stemcells/bosh-azure-hyperv-ubuntu-trusty-go_agent?v=3586.24"

**Deployment info:**

**Additional context**
There is no useful debug log got from cli, so I got the log of director (/var/vcap/sys/log @ director VM) and uploaded them to this [link](https://github.com/gossion/logs/tree/master/bosh-director-logs-for-task-stuck/director) in case you need more info.

**Comment by luan** [comment]
Hi @gossion,

We see some interesting looking errors in your `run.log` and the agent logs on the director that don't look normal to us:

```
2018-02-03_13:02:19.75449 panic: runtime error: invalid memory address or nil pointer dereference
2018-02-03_13:02:19.75450 [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x850e40]
2018-02-03_13:02:19.75450 
2018-02-03_13:02:19.75450 goroutine 9 [running]:
2018-02-03_13:02:19.75474 github.com/cloudfoundry/bosh-agent/infrastructure.(*MultiSourceMetadataService).GetPublicKey(0xc4201409f0, 0xc4200d1b80, 0xc57020, 0xc420010af0, 0xc5cbc0)
2018-02-03_13:02:19.75475 	/tmp/build/9674af12/gopath/src/github.com/cloudfoundry/bosh-agent/infrastructure/multi_source_metadata_service.go:17 +0x30
2018-02-03_13:02:19.75475 github.com/cloudfoundry/bosh-agent/infrastructure.ComplexSettingsSource.PublicSSHKeyForUsername(0xc5c200, 0xc4201409f0, 0xc513e0, 0xc420059020, 0x9d0b81, 0x15, 0xc5d280, 0xc420017a90, 0x9c22a4, 0x4, ...)
2018-02-03_13:02:19.75475 	/tmp/build/9674af12/gopath/src/github.com/cloudfoundry/bosh-agent/infrastructure/complex_settings_source.go:31 +0x31
2018-02-03_13:02:19.75475 github.com/cloudfoundry/bosh-agent/infrastructure.(*ComplexSettingsSource).PublicSSHKeyForUsername(0xc420142680, 0x9c22a4, 0x4, 0xc58660, 0xc420142480, 0x0, 0x0)
2018-02-03_13:02:19.75476 	<autogenerated>:10 +0x7c
2018-02-03_13:02:19.75500 github.com/cloudfoundry/bosh-agent/settings.(*settingsService).PublicSSHKeyForUsername(0xc42012b520, 0x9c22a4, 0x4, 0xc420140960, 0x25, 0xc420140960, 0x25)
2018-02-03_13:02:19.75501 	/tmp/build/9674af12/gopath/src/github.com/cloudfoundry/bosh-agent/settings/service.go:59 +0x4e
2018-02-03_13:02:19.75501 github.com/cloudfoundry/bosh-agent/agent.bootstrap.Run(0xc5f260, 0xc420142200, 0xc625e0, 0xc420083040, 0x9c6867, 0x9, 0xc59660, 0xc42012b520, 0xc5d280, 0xc420017a90, ...)
2018-02-03_13:02:19.75502 	/tmp/build/9674af12/gopath/src/github.com/cloudfoundry/bosh-agent/agent/bootstrap.go:50 +0x1b5
2018-02-03_13:02:19.75502 github.com/cloudfoundry/bosh-agent/agent.(*bootstrap).Run(0xc420010be0, 0xc420083040, 0x9c6867)
2018-02-03_13:02:19.75502 	<autogenerated>:16 +0x5c
2018-02-03_13:02:19.75502 github.com/cloudfoundry/bosh-agent/app.(*app).Setup(0xc42000ab00, 0x0, 0x0, 0x7ffeb6038f22, 0x6, 0x9c6867, 0x9, 0x9c2c3d, 0x5, 0x7ffeb6038f2c, ...)
2018-02-03_13:02:19.75504 	/tmp/build/9674af12/gopath/src/github.com/cloudfoundry/bosh-agent/app/app.go:112 +0xd2c
2018-02-03_13:02:19.75505 main.runAgent.func1(0xc5d280, 0xc420017a90, 0x0, 0x0, 0x7ffeb6038f22, 0x6, 0x9c6867, 0x9, 0x9c2c3d, 0x5, ...)
2018-02-03_13:02:19.75505 	/tmp/build/9674af12/gopath/src/github.com/cloudfoundry/bosh-agent/main/agent.go:30 +0x226
2018-02-03_13:02:19.75505 created by main.runAgent
2018-02-03_13:02:19.75505 	/tmp/build/9674af12/gopath/src/github.com/cloudfoundry/bosh-agent/main/agent.go:43 +0xb9
```

And 

```
[CLI] 2018/02/03 13:05:37 ERROR - Deploying: Creating instance 'bosh/0': Updating instance disks: Updating disks: Deploying disk: Mounting disk: Sending 'get_task' to the agent: Agent responded with error: Action Failed get_task: Task 0dd59011-9dfa-45c6-6f16-b6c046d87dfe result: Persistent disk with volume id 'caching:None;disk_name:bosh-disk-data-bae04939-b5dd-4330-b36e-8cc1e65505ef;resource_group_name:ICA-RG-ManagedCF-2-3-4-38-57' could not be found
```
Could you see in that director VM if the jobs are healthy and running? A `monit status` output would be helpful here.

**Comment by luan** [comment]
Sorry for accidentally closing and re-opening!

**Comment by gossion** [comment]
Thanks for response, @luan .

The `run.log` is for another issue, sorry for misleading. 
I have renamed folder in that github repo to avoid confusion. New folder is this case is https://github.com/gossion/logs/tree/master/issue-2043/, all logs in /var/vcap/sys/log are uploaded, and /var/vcap/monit/ is also uploaded to this folder.

For this case, we did not have a `run.log` (`BOSH_LOG_PATH` was not set). 

`monit summary` shows that everything is running, see logs as below:
```
azureuser@devbox-8-28-7-34-11:~$ bosh -e azure tasks -r
Using environment '10.0.0.4' as client 'admin'

ID  State   Started At                    Last Activity At              User   Deployment  Description      Result
7   queued  Thu Jan  1 00:00:00 UTC 1970  Tue Aug 28 15:08:54 UTC 2018  admin  -           create release   -
6   done    Tue Aug 28 15:09:17 UTC 2018  Tue Aug 28 15:09:37 UTC 2018  admin  -           create release   Created release 'bpm/0.9.0'
5   done    Tue Aug 28 15:08:25 UTC 2018  Tue Aug 28 15:09:18 UTC 2018  admin  -           create release   Created release 'silk/2.10.0'
4   done    Tue Aug 28 15:09:18 UTC 2018  Tue Aug 28 15:09:21 UTC 2018  admin  -           create release   Created release 'bosh-dns-aliases/0.0.2'
3   done    Tue Aug 28 15:08:25 UTC 2018  Tue Aug 28 15:09:17 UTC 2018  admin  -           create release   Created release 'loggregator-agent/2.0'
2   done    Tue Aug 28 15:04:57 UTC 2018  Tue Aug 28 15:17:19 UTC 2018  admin  -           create stemcell  /stemcells/bosh-azure-hyperv-ubuntu-trusty-go_agent/3586.36
1   done    Tue Aug 28 15:04:39 UTC 2018  Tue Aug 28 15:04:56 UTC 2018  admin  -           create release   Created release 'bosh-dns/1.7.0'

7 tasks

Succeeded


azureuser@devbox-8-28-7-34-11:~$ ./connect_director_vm.sh
Unauthorized use is strictly prohibited. All access and activity
is subject to logging and monitoring.
Welcome to Ubuntu 14.04.5 LTS (GNU/Linux 4.4.0-128-generic x86_64)

 * Documentation:  https://help.ubuntu.com/
Last login: Wed Aug 29 07:37:16 UTC 2018 from 10.0.0.100 on pts/0
Last login: Thu Aug 30 01:36:19 2018 from 10.0.0.100
bosh/0:~$ sudo su
bosh/0:/home/jumpbox# monit summary
The Monit daemon 5.2.5 uptime: 1d 10h 36m

Process 'nats'                      running
Process 'postgres'                  running
Process 'blobstore_nginx'           running
Process 'director'                  running
Process 'worker_1'                  running
Process 'worker_2'                  running
Process 'worker_3'                  running
Process 'worker_4'                  running
Process 'director_scheduler'        running
Process 'director_sync_dns'         running
Process 'director_nginx'            running
Process 'health_monitor'            running
Process 'registry'                  running
Process 'uaa'                       running
Process 'credhub'                   running
System 'system_localhost'           running


bosh/0:/home/jumpbox# monit status
The Monit daemon 5.2.5 uptime: 1d 10h 36m

Process 'nats'
  status                            running
  monitoring status                 monitored
  pid                               36883
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  7932
  memory kilobytes total            7932
  memory percent                    0.1%
  memory percent total              0.1%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'postgres'
  status                            running
  monitoring status                 monitored
  pid                               36922
  parent pid                        1
  uptime                            1d 10h 35m
  children                          34
  memory kilobytes                  13760
  memory kilobytes total            357140
  memory percent                    0.1%
  memory percent total              5.0%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'blobstore_nginx'
  status                            running
  monitoring status                 monitored
  pid                               36984
  parent pid                        1
  uptime                            1d 10h 35m
  children                          2
  memory kilobytes                  4852
  memory kilobytes total            20612
  memory percent                    0.0%
  memory percent total              0.2%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'director'
  status                            running
  monitoring status                 monitored
  pid                               37029
  parent pid                        1
  uptime                            1d 10h 35m
  children                          3
  memory kilobytes                  53916
  memory kilobytes total            252624
  memory percent                    0.7%
  memory percent total              3.5%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'worker_1'
  status                            running
  monitoring status                 monitored
  pid                               38186
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  57956
  memory kilobytes total            57956
  memory percent                    0.8%
  memory percent total              0.8%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'worker_2'
  status                            running
  monitoring status                 monitored
  pid                               38203
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  58468
  memory kilobytes total            58468
  memory percent                    0.8%
  memory percent total              0.8%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'worker_3'
  status                            running
  monitoring status                 monitored
  pid                               38217
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  58348
  memory kilobytes total            58348
  memory percent                    0.8%
  memory percent total              0.8%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'worker_4'
  status                            running
  monitoring status                 monitored
  pid                               38236
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  58760
  memory kilobytes total            58760
  memory percent                    0.8%
  memory percent total              0.8%
  cpu percent                       0.4%
  cpu percent total                 0.4%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'director_scheduler'
  status                            running
  monitoring status                 monitored
  pid                               37126
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  64544
  memory kilobytes total            64544
  memory percent                    0.9%
  memory percent total              0.9%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'director_sync_dns'
  status                            running
  monitoring status                 monitored
  pid                               37167
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  64988
  memory kilobytes total            64988
  memory percent                    0.9%
  memory percent total              0.9%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'director_nginx'
  status                            running
  monitoring status                 monitored
  pid                               37193
  parent pid                        1
  uptime                            1d 10h 35m
  children                          2
  memory kilobytes                  4604
  memory kilobytes total            20488
  memory percent                    0.0%
  memory percent total              0.2%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'health_monitor'
  status                            running
  monitoring status                 monitored
  pid                               40348
  parent pid                        1
  uptime                            1d 10h 19m
  children                          0
  memory kilobytes                  37968
  memory kilobytes total            37968
  memory percent                    0.5%
  memory percent total              0.5%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  port response time                0.001s to localhost:25923/healthz [HTTP via TCP]
  data collected                    Thu Aug 30 01:36:26 2018

Process 'registry'
  status                            running
  monitoring status                 monitored
  pid                               37286
  parent pid                        1
  uptime                            1d 10h 35m
  children                          0
  memory kilobytes                  31920
  memory kilobytes total            31920
  memory percent                    0.4%
  memory percent total              0.4%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

Process 'uaa'
  status                            running
  monitoring status                 monitored
  pid                               37320
  parent pid                        1
  uptime                            1d 10h 35m
  children                          10
  memory kilobytes                  415456
  memory kilobytes total            430332
  memory percent                    5.8%
  memory percent total              6.0%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  port response time                0.000s to localhost:8443/healthz [HTTP via TCP]
  data collected                    Thu Aug 30 01:36:26 2018

Process 'credhub'
  status                            running
  monitoring status                 monitored
  pid                               38398
  parent pid                        1
  uptime                            1d 10h 34m
  children                          12
  memory kilobytes                  630636
  memory kilobytes total            645668
  memory percent                    8.8%
  memory percent total              9.0%
  cpu percent                       0.0%
  cpu percent total                 0.0%
  data collected                    Thu Aug 30 01:36:26 2018

System 'system_localhost'
  status                            running
  monitoring status                 monitored
  load average                      [0.06] [0.12] [0.05]
  cpu                               0.8%us 0.3%sy 0.8%wa
  memory usage                      1972724 kB [27.6%]
  swap usage                        30028 kB [0.4%]
  data collected                    Thu Aug 30 01:36:26 2018

bosh/0:/home/jumpbox#
```

The error happens several times recently, but is not always reproduced. Last time we saw the issue, we redeployed the director, and then the issue was gone.

**Comment by jfmyers9** [comment]
Hey @gossion,

Seems like the workers aren't picking up the queued delayed jobs. I believe I have seen this before, and just restarting the worker processes seems to unblock this.

Is there any chance you could give us the contents of the `delayed_jobs` table? This might give us some more insight into what is currently queued. This can be done by running:

```
class J < ::Sequel::Model(:delayed_jobs) ; end
J.all
```
in the director console, or by using the `psql` binary to list the contents of the `delayed_jobs` table.

Thanks,

@jfmyers9 && @mikexuu, CF BOSH

**Comment by gossion** [comment]
Thanks @jfmyers9 . 
Unfortunately, the test environment was deleted. I will keep monitoring the error, and get `delayed_jobs` for you if it happens again.

**Comment by s4heid** [comment]
We also observed this issue today. After executing

```
class J < ::Sequel::Model(:delayed_jobs) ; end
J.all
```

we saw that the `delayed jobs` table was empty even though the director showed 170 queued tasks.

The reason was that the persistent disk was full when the tasks were created which caused this anomaly.

We were able to fix it by increasing the disk size and cancelling the queued tasks with:
`bosh cancel-tasks`.

cc @BeckerMax

**Comment by jvshahid** [comment]
We ran into the same thing after the director's persistent disk became full over the weekend.  This is the output from the director console:

```
 irb(main):001:0> class J < ::Sequel::Model(:delayed_jobs) ; end
=> nil
irb(main):002:0> J.all
=> []
irb(main):003:0>

```
